AndroidMediaCodec: Fallback to HEVC if Dolby Vision is not fully supp…

quietvoid opened the pull request:
…rted

## Description
Allow Dolby Vision playback through MediaCodec only if the playback chain fully supports it.
This assumes minimum Dolby Vision compatibility, which is usually profile 5 on Android devices.

## Motivation and context
Avoid falling back to a software decoder in case the Android device doesn't support Dolby Vision.

This is mostly to avoid breaking compatibility when the hinted codec_tag is `dvhe`/`dvh1`, etc.
Otherwise, the mime type tested would always be `video/dolby-vision`, without ever trying for regular HEVC hardware acceleration.

## How has this been tested?
Tested on a FireTV Stick 4K Max.
With HDR adaptive enabled, both display and mediacodec support Dolby Vision.
When disabling HDR, only the display compatibility is forced to false by the device.

On a 2019 FireTV Stick connected to a standard HDTV:
`CDVDVideoCodecAndroidMediaCodec::Open Dolby Vision support: Display false, MediaCodec: false`
And it correctly picks the hardware HEVC decoder (though it fails configuring because it doesn't support 10 bit.)

## What is the effect on users?
Should avoid problems related to HDR10 fallback for non Dolby Vision setups, with no negative effect.

Currently, only inputstream.adaptive is using the hints so the change shouldn't do much if anything at all.
I can't get the `video/dolby-vision` mimetype at all without an extra patch for local files/strms, if the CodecID is not already `dvhe`.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

quietvoid commented:
I'm not sure if `CAndroidUtils` is the right place for the helper function, since it seems to be focused on windowing.

ksooo commented on xbmc/windowing/android/AndroidUtils.h:
```diff
@@ -31,6 +31,8 @@ class CAndroidUtils : public ISettingCallback
   static const std::string SETTING_LIMITGUI;
   void OnSettingChanged(const std::shared_ptr<const CSetting>& setting) override;
 
+  static bool SupportsMediaCodecMimeType(const std::string mimeType);
```
`const std::string&`

ksooo commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp:
```diff
@@ -31,6 +31,8 @@
 #include "utils/TimeUtils.h"
 #include "utils/log.h"
+#include "windowing/android/AndroidUtils.h"
+#include "windowing/WinSystem.h"
```
Making the codec dependent on 'windowing' feels wrong. What has a codec to do with the window system?

quietvoid replied:
It doesn't, I just don't know where else this can be implemented.

quietvoid replied:
It should probably be all passed as a stream hint, I guess.

ksooo replied:
> include is required for CServiceBroker::GetWinSystem()->GetDisplayHDRCapabilities().

This information originally comes from the Android specific code we are in here, can’t you get the info directly from that low level layer?

quietvoid replied:
@ksooo Do you mean using `CAndroidUtils::GetDisplaySupportedHdrTypes` directly? I guess I could, but it's not as clean.

a1rwulf replied:
@quietvoid what do you mean with it's an instance function?

quietvoid replied:
I might be mistaken, but I think the `CAndroidUtils::GetDisplaySupportedHdrTypes()` function can only be called on an instance of the class?
Although it doesn't use the member variables at all.

This is also why I made the new `CAndroidUtils::SupportsMediaCodecMimeType` function static.

a1rwulf replied:
I'm not an expert in the area, but if it doesn't use any members, what's the issue with turning it into a static method instead?

quietvoid replied:
I think it would work, I just don't know if it's correct.

a1rwulf replied:
If I look at this class, I don't see any reason why we would need to make any method non-static that does not access one of the members.
@ksooo What's your opinion on this?

quietvoid replied:
I've updated the ones I'm using to be static, it seems to work fine.

thexai replied:
> The WinSystem.h include is required for CServiceBroker::GetWinSystem()->GetDisplayHDRCapabilities()

An alternative way can be:

```c++
  auto winSystem = dynamic_cast<CWinSystemAndroid*>(CServiceBroker::GetWinSystem());
  winSystem->GetDisplayHDRCapabilities()
```

quietvoid replied:
@thexai In the latest changes I've just moved the Android method to `CAndroidUtils` so the win system isn't being used.

thexai commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp:
```diff
@@ -521,6 +523,23 @@ bool CDVDVideoCodecAndroidMediaCodec::Open(CDVDStreamInfo &hints, CDVDCodecOptio
+        CLog::Log(LOGDEBUG, "CDVDVideoCodecAndroidMediaCodec::Open Dolby Vision support: "
+                            "Display {}, MediaCodec: {}", displaySupportsDovi, mediaCodecSupportsDovi);
+
+        if (!displaySupportsDovi || !mediaCodecSupportsDovi)
```
It seems to me as if the house had begun to be built from the roof and the order is illogical.

I mean: 

by default mime = video/hevc
If the system supports DV then mime = video/dolby-vision

something like this:

```c++
case AV_CODEC_ID_HEVC:
  
  m_mime = "video/hevc";
  m_formatname = "amc-hevc";

  const bool isDvhe = (m_hints.codec_tag == MKTAG('d', 'v', 'h', 'e'));
  const bool isDvh1 = (m_hints.codec_tag == MKTAG('d', 'v', 'h', '1'));

  if (isDvhe || isDvh1)
  {
    const bool displaySupportsDovi = CServiceBroker::GetWinSystem()->GetDisplayHDRCapabilities().SupportsDolbyVision();
    const bool mediaCodecSupportsDovi = CAndroidUtils::SupportsMediaCodecMimeType("video/dolby-vision");
      
    if (displaySupportsDovi && mediaCodecSupportsDovi)
    {
      if (isDvhe)
      {
        m_mime = "video/dolby-vision";
        m_formatname = "amc-dvhe";
      }
      else if (isDvh1)
      {
        m_mime = "video/dolby-vision";
        m_formatname = "amc-dvh1";
      }
    }
  }

. . . 

break;
```

quietvoid replied:
Thanks, that indeed does make a lot more sense.

thexai commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp:
```diff
@@ -500,27 +501,41 @@ bool CDVDVideoCodecAndroidMediaCodec::Open(CDVDStreamInfo &hints, CDVDCodecOptio
+        {
+          m_mime = "video/dolby-vision";
+
+          if (isDvhe)
```
Good place for ternary operator 👌

`m_formatname = isDvhe ? "amc-dvhe" : "amc-dvh1";`

quietvoid commented:
There's no more dependency on `WinSystem` anymore.
I just need to fix the log consistency and check if it builds.

quietvoid committed:
AndroidMediaCodec: Fallback to HEVC if Dolby Vision is not fully supported

Allow Dolby Vision playback through MediaCodec only if the playback chain fully supports it.
This assumes minimum Dolby Vision compatibility, which is usually profile 5 on Android devices.

Also refactor CAndroidUtils a bit to make methods static because they don't use member variables.

thexai commented:
I'm going to close/open PR to fix jenkins issues

thexai commented:
@quietvoid
Have you been able to test fallback to HDR10 when using a stream HDR10 + DoVi compatible?

quietvoid commented:
> @quietvoid
> Have you been able to test fallback to HDR10 when using a stream HDR10 + DoVi compatible?

I don't own a HDR10 only device nor display, so I have not.
I think it would require testing on a 2015/2017 Shield TV.

I don't see why it wouldn't fallback properly, though.

thexai commented:
It was just to get confirmation, I also think it should work fine.

I think it could also be checked in Shield TV 2019 since it has a setting to disable DV only, according to this capture:

![dv](https://user-images.githubusercontent.com/58434170/142684531-098f3f53-6af2-4b1a-861e-4e1817e654ff.png)

quietvoid commented:
Tried playback with the Netflix addon and HDR disabled and the correct behaviour happens.
The HEVC decoder is selected instead of DVHE.

quietvoid commented:
Somehow installing a build with this patch on my 2018 4K stick broke the system info HDR supported types, where it always says None.
But on playback the debug logs say that the display is HDR and Dolby Vision is supported.

a1rwulf commented:
@quietvoid I can give it a shot, do you have a testfile for me?

quietvoid commented:
@a1rwulf Sure, here's one: https://mega.nz/file/4F9gib7B#VLH2fdxvEQfFYysdOI_jU30qNvBEVib46zvEWjY5FFM
It's the Dolby Vision sample profile 8.1 stream, but remuxed to have `codec_tag=dvhe`.

a1rwulf commented:
Can't find the jenkins artefact.
Jenkins build this please.

a1rwulf commented:
Alright.
I've tested this PR with following results:
First, I don't have the menu in the screenshot above on my Shield Pro.
However I have the following modes to choose:
![IMG_20211122_192051](https://user-images.githubusercontent.com/5141257/142916583-b4c409ea-0d1c-46bd-860f-cbb446189fe5.jpg)

If I choose the Dolby Vision enabled mode my TV goes into Dolby Vision mode:
https://paste.kodi.tv/etukolunek.kodi

If I choose the default mode my TV goes into HDR10 mode:
https://paste.kodi.tv/ovucakoleb.kodi

Guess this is what you'd expect from that PR :-)

a1rwulf reviewed (approved):
Imho code looks good and I've successfully tested the fallback on my setup.

quietvoid commented:
Looks correct according to the log too. Thanks for testing.

thexai commented:
Thank you for yet another DoVi related improvement!
